### ISSUE TITLE
Remove plenary dependency

### ISSUE THREAD

**Issue by Conni2461** [nvim-telescope/telescope.nvim#2552]
and replace with vim internals (https://github.com/nvim-telescope/telescope.nvim/pull/2529#issuecomment-1580324518) or maybe we want to slim down plenary

**Comment by clason** [comment]
I think it makes sense to tackle this here one module at a time; replacing a single stale plenary module with a maintained upstream module is a win itself, even if plenary itself remains a dependency.

I should also mention that it's perfectly fine to make feature requests (or PRs) to Neovim if one of the builtin modules is missing something that's needed here (and not too telescope-specific; otherwise just vendoring it here is best). 

Of course, slimming down plenary and removing obsolete modules (or refactoring them based on upstream) is appreciated as well! I've heard repeatedly that the (understandable) lack of actual development there has dissuaded someone from contributing there. Something to consider is also to more strongly modularize plenary along the lines of https://github.com/echasnovski/mini.nvim (so each module can be installed independently as well).

From a brief grep of telescope, here's what's used, with possible replacements:

* [ ] path -> vim.fs (although that's more targeted; a `vim.paths` module may be welcome)
* [ ] job -> vim.system (on master/0.10)
* [ ] strings (a `vim.strings` module may be welcome)
* [ ] scandir -> vim.fs
* [x] filetype -> vim.filetype
* [x] context_manager
* [ ] class 
* [ ] log (if Neovim's logging functionality is missing something, consider opening an issue)
* [ ] async -> vim.async (planned), or can be replaced with a vendored minimal module as in https://github.com/nvim-treesitter/nvim-treesitter/blob/main/lua/nvim-treesitter/async.lua
* [ ] popup (maybe https://github.com/MunifTanjim/nui.nvim?)
* [ ] plenary.busted / plenary.test_harness -> https://github.com/lewis6991/nvim-test (once https://github.com/lewis6991/nvim-test/pull/12 is merged)

**Comment by Conni2461** [comment]
[ ] `plenary.busted` / `plenary.test_harness` -> ?

**Comment by clason** [comment]
Das habe ich jetzt nicht gezählt, weil das keine dependency für telescope selbst ist -- das kann man rein im Rahmen von CI handhaben (so machen wir das bei nvim-treesitter auch).

**Comment by clason** [comment]
> I need to evaluate every module because i'm no longer that much up to date with neovim development 

Of course; This Is The Way. Let me know if you have any questions about the Neovim side (here or on Matrix).

**Comment by max397574** [comment]
I think plenary.strings could be quite easily upstreamed
it's nearly standalone
the only thing that's required from other modules is the [`path.sep`](https://github.com/nvim-lua/plenary.nvim/blob/499e0743cf5e8075cd32af68baa3946a1c76adf1/lua/plenary/path.lua#L21-L32)
which imo could also be upstreamed since it's something you see really often in plugins and configs

**Comment by clason** [comment]
I don't think `plenary.strings` is upstreamable as-is. The way these things are done is that we _first_ identify what we _want_ to have in core -- including the API -- and then implement it. Requirements are often quite different for core than for a plugin.

(In particular, as @Conni2461 wrote, core should make use of Neovim's string handling functions in C as much as possible.)

**Comment by tjdevries** [comment]
my only concern with the way things are for vim.system right now is I don't think it's as performant for doing line based items (which a lot of operations are for neovim). One thing that would be really awesome for vim.system would be some `on_line` or `lines` parameter or similar, and maybe some string wrangling in C to make things really zoom. But not sure, I'd have to do some benchmarks again.

**Comment by tjdevries** [comment]
Would have really liked `popup` to have been fully fleshed out and made to match some of the vim apis, since that's a nice two-for-one kind of situation where a bunch of plugins can support vim and nvim with the same functions. But that doesn't look like it's going to happen (haven't seen any progress on that and I don't have any desire to drive those kinds of projects at the moment)

**Comment by tjdevries** [comment]
Nice to see that maybe plenary can just be deprecated soon and a lot of the stuff has been upstreamed (and done better)

**Comment by clason** [comment]
> my only concern with the way things are for vim.system right now is I don't think it's as performant for doing line based items (which a lot of operations are for neovim). One thing that would be really awesome for vim.system would be some `on_line` or `lines` parameter or similar, and maybe some string wrangling in C to make things really zoom. But not sure, I'd have to do some benchmarks again.

That's not a bad idea; we have `vim.iter` now, and being able to feed `stdout` into that would be quite neat.

(I've also had situations where I'd have preferred a `systemlist`-type table output, but on synchronous jobs that's just a `vim.split` away, of course.)

**Comment by tjdevries** [comment]
@lewis6991 from what I could tell, it always tries to put all the stdout into one variable at the end of execution? I have to go look again.

My main concern is when you do something like `rg --files` from home directory, i don't actually want to try and concat the string and save that all in memory (duplicated) again.

I can take a look again later tho.

**Comment by lewis6991** [comment]
It only does that if you **don't** provide your own handler. If you provide a handler of the form:

```lua
function stdou_handler(err, data)
end
```

That will be used instead of collecting the data into a variable.

**Comment by tjdevries** [comment]
ah, nice. thanks

**Comment by clason** [comment]
> Would have really liked `popup` to have been fully fleshed out and made to match some of the vim apis, since that's a nice two-for-one kind of situation where a bunch of plugins can support vim and nvim with the same functions. But that doesn't look like it's going to happen (haven't seen any progress on that and I don't have any desire to drive those kinds of projects at the moment)

I think that project has become a victim of Neovim's success: before the post-0.5 Cambrian Explosion of Lua plugins, enticing Vimscript plugin devs with such an API seemed like a great way of growing the Neovim ecosystem. Now I could count the number of plugins profiting from this on one (saw-mill worker's) hand since they'd have to 
1. be originally written for Vim only;
2. make heavy use of popup UI;
3. are established and don't have any Lua peers;
4. don't have full Neovim-specific alternate code paths for this already.

As it stands, it would make much more sense to add a Neovim-specific high(er) level `vim.ui.popup()` API in core (designed fresh with all the lessons learned).

**Comment by tjdevries** [comment]
Yeah, but it does make situations like frameworks (denops comes to mind) where you want to have some shared API that works well in both quite nice.

(I just wanted to implement the vimscript APIs in Lua basically as wrappers around our code. A lot of it is quite nice for simple stuff that people generally do). 

Yeah, it's probably not a huge win tho

**Comment by juanandresgs** [comment]
Updated plugins today and got the following error
<img width="1081" alt="image" src="https://github.com/nvim-telescope/telescope.nvim/assets/15952691/a7504f36-e645-40f4-820e-660dab95c3ac">
Guessing it's related to this discussion on frencency/plenary?

```NVIM v0.9.2 (LunarVim 1.3)
Build type: Release
LuaJIT 2.1.0-beta3```

**Comment by clason** [comment]
Nope.

**Comment by delphinus** [comment]
@juanandresgs This is a bug by telescope-frecency.nvim. I fixed. Check it.

https://github.com/nvim-telescope/telescope-frecency.nvim/pull/147

**Comment by mrcjkb** [comment]
> * `plenary.busted` / `plenary.test_harness`

With `nvim -l` it's quite easy to use [`busted`](https://github.com/lunarmodules/busted) with Neovim as the interpreter, without plenary. See [my blog post on the topic for details](https://mrcjkb.dev/posts/2023-06-06-luarocks-test.html).

Some examples where this has been done:

- https://github.com/mrcjkb/haskell-tools.nvim/blob/d9d15b6754aa28373eb93876553d7e429f524374/nix/test-overlay.nix#L56
- https://github.com/lewis6991/gitsigns.nvim/blob/bce4576a9047085a528c479a7fe1e2f6b787b6c1/Makefile#L51
- The [luarocks-tag-release GitHub action](https://github.com/nvim-neorocks/luarocks-tag-release) runs busted tests (with neovim) if it finds a `.busted` file in the project root (without releasing on PR).

**Comment by corsinmusic** [comment]
Plenary will be archived soon, btw: https://github.com/nvim-lua/plenary.nvim

> ## Important Notice
> This repository is no longer actively maintained and will be officially archived soon.
> - Critical bugs may still be addressed until 2026-06-30
> - No new features will be added
> - Ongoing support should not be expected

**Comment by Conni2461** [comment]
thanks but we've decided to archive it :D we are working on this in #3647
